// Copyright Robert Bosch GmbH, 2022. Part of the Eclipse Kuksa Project.
//
// All rights reserved. This configuration file is provided to you under the
// terms and conditions of the Eclipse Distribution License v1.0 which
// accompanies this distribution, and is available at
// http://www.eclipse.org/org/documents/edl-v10.php

// This is a base proto file for databroker and kuksa-val-basic
// function set

syntax = "proto3";

package kuksa.val.v1;

import "kuksa/val/v1/types.proto";

// Should probably be changed to something like:
// option go_package = "github.com/eclipse/kuksa.val/proto/v1";
option go_package = "/kuksa_grpc_proto";

// Note on authorization:
// Tokens (auth-token or auth-uuid) are sent as (GRPC / http2) metadata.
//
// The auth-token is a JWT compliant token as the examples found here:
// https://github.com/eclipse/kuksa.val/tree/master/kuksa_certificates/jwt
//
// See also https://github.com/eclipse/kuksa.val/blob/master/doc/jwt.md
//
// Upon reception of auth-token, server shall generate an auth-uuid in metadata
// that the client can use instead of auth-token in subsequent calls.

service VAL {
  //  Get datapoints
  rpc Get(GetRequest) returns (GetResponse);

  // Set datapoints
  rpc Set(SetRequest) returns (SetResponse);

  // Subscribe to a set of data points
  //
  // Returns a stream of replies.
  //
  // InvalidArgument is returned if the request is malformed.
  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse);

  // Shall return information that allows the client to determine
  // what server/server implementation/version it is talking to
  // eg. kuksa-databroker 0.5.1
  rpc GetServerInfo(GetServerInfoRequest) returns (GetServerInfoResponse);
}

// Describes which data we want
message GetDatapoint {
  string path  = 1;
  View view    = 2;
  Field fields = 3;
}

// Request a number of VSS datapoints.
message GetRequest {
  repeated GetDatapoint datapoints = 1;
}

// Global errors are specified in `error`.
// Errors for individual datapoints are specified in `errors`.
message GetResponse {
  repeated Datapoint datapoints  = 1;
  repeated DatapointError errors = 2;
  Error error                    = 3;
}

// Describes which data we want to set
message SetDatapoint {
  string path           = 1;
  Datapoint datapoint   = 2;
  repeated Field fields = 3;
}

// A list of datapoints to be set
message SetRequest {
  repeated SetDatapoint datapoints = 1;
}

// Global errors are specified in `error`.
// Errors for individual datapoints are specified in `errors`.
message SetResponse {
  Error error                              = 1;
  repeated DatapointError datapoint_errors = 2;
}

// Describes what to subscribe to
message SubscribeDatapoint {
  string path           = 1;
  View view             = 2;
  repeated Field fields = 3;
}

// Subscribe to changes in datapoints.
message SubscribeRequest {
  repeated SubscribeDatapoint datapoints = 1;
}

// The updated datapoint
// `fields` specify which fields where changed.
message UpdatedDatapoint {
  Datapoint datapoint   = 1;
  repeated Field fields = 2;
}

// A subscription response
message SubscribeResponse {
  repeated UpdatedDatapoint datapoints = 1;
}

message GetServerInfoRequest {
  // Nothing yet
}

message GetServerInfoResponse {
  string name    = 1;
  string version = 2;
}